failure in test-tls13-finished.py

# Bug Report

The test `cipher TLS_AES_256_GCM_SHA384, pad_byte 0, pad_left 0, pad_right 16777167` in `test-tls13-finished.py` failed in Travis build.

## System Information

| Type                     | Version/Name |
| ------------------------ | ------------ |
| Operating System Name    | <!-- provide name of OS you're using here, e.g.: Windows, RHEL --> |
| Operating System Version | <!-- provide the version of the OS you're using here, e.g.: 6.10, 7 --> |
| Python version           | python 3.5 |
| tlslite-ng version       | tlslite-ng>=0.8.0-alpha16 |
| tlsfuzzer revision       | <!-- provide the revision of tlsfuzzer that you are using (use `git rev-parse HEAD` in the checked-out tlsfuzzer directory --> |

## Problem description
Build fails in script regression testing, test-tls13-finished.py

## Describe how to reproduce the problem
run the test mutiple times in travis

## Expected behaviour
no failures

## Include errors or backtraces

```
test-tls13-finished.py:stdout:padding - cipher TLS_AES_256_GCM_SHA384, pad_byte 0, pad_left 0, pad_right 16777167 ...
server:stdout:About to handshake...
test-tls13-finished.py:stdout:Error encountered while processing node <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f0ec9dbbc88> (child: <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f0ec9dbbc88>) with last message being: None
test-tls13-finished.py:stdout:Error while processing
test-tls13-finished.py:stdout:Traceback (most recent call last):
test-tls13-finished.py:stdout:  File "/home/travis/build/tomato42/tlsfuzzer/tlsfuzzer/runner.py", line 180, in run
test-tls13-finished.py:stdout:    header, parser = self.state.msg_sock.recvMessageBlocking()
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/messagesocket.py", line 101, in recvMessageBlocking
test-tls13-finished.py:stdout:    for res in self.recvMessage():
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/messagesocket.py", line 83, in recvMessage
test-tls13-finished.py:stdout:    for ret in self.recvRecord():
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/recordlayer.py", line 860, in recvRecord
test-tls13-finished.py:stdout:    for result in self._recordSocket.recv():
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/recordlayer.py", line 201, in recv
test-tls13-finished.py:stdout:    for record in self._recvHeader():
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/recordlayer.py", line 135, in _recvHeader
test-tls13-finished.py:stdout:    for result in self._sockRecvAll(1):
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/recordlayer.py", line 112, in _sockRecvAll
test-tls13-finished.py:stdout:    socketBytes = self.sock.recv(length - len(buf))
test-tls13-finished.py:stdout:  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/tlslite/bufferedsocket.py", line 54, in recv
test-tls13-finished.py:stdout:    return self.socket.recv(bufsize)
test-tls13-finished.py:stdout:socket.timeout: timed out
test-tls13-finished.py:stdout:
test-tls13-finished.py:stdout:During handling of the above exception, another exception occurred:
test-tls13-finished.py:stdout:
test-tls13-finished.py:stdout:Traceback (most recent call last):
test-tls13-finished.py:stdout:  File "scripts/test-tls13-finished.py", line 431, in main
test-tls13-finished.py:stdout:    runner.run()
test-tls13-finished.py:stdout:  File "/home/travis/build/tomato42/tlsfuzzer/tlsfuzzer/runner.py", line 196, in run
test-tls13-finished.py:stdout:    raise AssertionError("Unexpected closure from peer")
test-tls13-finished.py:stdout:AssertionError: Unexpected closure from peer
test-tls13-finished.py:stdout:
```

## Additional context
build - https://travis-ci.org/tomato42/tlsfuzzer/builds/408919292

--- a/scripts/test-tls13-finished.py
+++ b/scripts/test-tls13-finished.py
@@ -28,7 +28,7 @@ from tlslite.extensions import KeyShareEntry, ClientKeyShareExtension, \
 from tlsfuzzer.helpers import RSA_SIG_ALL, key_share_ext_gen
 
 
-version = 2
+version = 3
 
 
 def help_msg():
@@ -335,8 +335,22 @@ def main():
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 8, 8) # SHA-512 size
         ]
     for cipher, pad_byte, pad_left, pad_right in scenarios:
-        # longer timeout for 16M messages
-        timeout = 5 if max(pad_left, pad_right) < 2**14 else 30
+        # longer timeout for longer messages
+        # Because the client is sending encrypted data without waiting
+        # on any server response, it can actually produce data at a faster
+        # rate than the server is able to process it, meaning that a server
+        # that aborts only after decrypting a full handshake message may have
+        # quite a few records in the queue after we, as a client have finished
+        # sending them. Since tlslite-ng has the ciphers implemented in
+        # pure python, they are very slow, speeds of just 71.5KiB/s for
+        # AES-256-GCM are not atypical. which translates to about 4 minutes
+        # to transfer this data. Set the timeout to 5 for a small margin of
+        # error.
+        # Note: because we still are waiting for the server to send us an alert
+        # (all graph terminal nodes go through ExpectAlert), server that fails
+        # to do that will still cause the whole test conversation to fail in
+        # case it just closes the connection on us
+        timeout = 5 if max(pad_left, pad_right) < 2**14 * 4 else 300
         conversation = Connect(host, port, timeout=timeout)
         node = conversation
         ciphers = [cipher,

--- a/tests/test_tlsfuzzer_runner.py
+++ b/tests/test_tlsfuzzer_runner.py
@@ -217,8 +217,27 @@ class TestRunner(unittest.TestCase):
         runner.state.msg_sock.recvMessageBlocking = \
                 mock.MagicMock(side_effect=TLSAbruptCloseError())
 
-        with self.assertRaises(AssertionError):
+        with self.assertRaises(AssertionError) as e:
             runner.run()
+
+        self.assertIn("Unexpected closure from peer", str(e.exception))
+
+    def test_run_with_expect_and_read_timeout(self):
+        node = mock.MagicMock()
+        node.is_command = mock.Mock(return_value=False)
+        node.is_expect = mock.Mock(return_value=True)
+        node.is_generator = mock.Mock(return_value=False)
+        node.child = None
+
+        runner = Runner(node)
+        runner.state.msg_sock = mock.MagicMock()
+        runner.state.msg_sock.recvMessageBlocking = \
+                mock.MagicMock(side_effect=socket.timeout())
+
+        with self.assertRaises(AssertionError) as e:
+            runner.run()
+
+        self.assertIn("Timeout when waiting", str(e.exception))
 
     def test_run_with_expect_and_no_message(self):
         node = ExpectNoMessage()

--- a/tlsfuzzer/runner.py
+++ b/tlsfuzzer/runner.py
@@ -193,7 +193,12 @@ class Runner(object):
                             node = close_node.child
                             continue
                         else:
-                            raise AssertionError("Unexpected closure from peer")
+                            if isinstance(exc, socket.timeout):
+                                raise AssertionError(
+                                    "Timeout when waiting for peer message")
+                            else:
+                                raise AssertionError(
+                                    "Unexpected closure from peer")
                     msg = Message(header.type, parser.bytes)
                     old_node = node
 

--- a/tlsfuzzer/runner.py
+++ b/tlsfuzzer/runner.py
@@ -177,7 +177,8 @@ class Runner(object):
                         self.state.msg_sock.sock.settimeout(node.timeout)
                     # check peer response
                     try:
-                        header, parser = self.state.msg_sock.recvMessageBlocking()
+                        header, parser = self.state.msg_sock.\
+                            recvMessageBlocking()
                     except (TLSAbruptCloseError, socket.error) as exc:
                         if isinstance(exc, socket.timeout) and \
                                 isinstance(node, ExpectNoMessage):
@@ -186,8 +187,9 @@ class Runner(object):
                             self.state.msg_sock.sock.settimeout(old_timeout)
                             node = node.child
                             continue
-                        close_node = next((n for n in node.get_all_siblings() \
-                                           if isinstance(n, ExpectClose)), None)
+                        close_node = next((n for n in node.get_all_siblings()
+                                           if isinstance(n, ExpectClose)),
+                                          None)
                         if close_node:
                             close_node.process(self.state, None)
                             node = close_node.child

